File: python/expmax.py
"""
An implementation of the exp-max algorithm.
Author: Tom Shen
Date: 2013/02/01
"""
from math import fabs
from random import randint
from scipy.stats import norm
import numpy as np
from kgauss import kgauss, kgauss_with_mus, kgauss_with_mus_sigmas
from util import timed, importFile, exportFile

def multi_expectation_maximization(data, dim, k, sigmas):
    assert(len(sigmas) == k)
    if dim == 1:
        return expectation_maximization(data, k, sigmas)
    h = []
    for d in xrange(dim):
        h.append(expectation_maximization(data[d], k, sigmas[d]))
    return h

def expectation_maximization(data, k, sigmas):
    mus = initial_hypothesis_mu(data, k)
    mus_old = None
    sigmas_old = None
    while compare_hypothesis(mus, mus_old) or compare_hypothesis(sigmas, sigmas_old):
        exp_val = calculate_expectation(k, data, mus, sigmas)
        mus_old = mus
        sigmas_old = sigmas
        mus, sigmas = calculate_hypothesis(k, data, exp_val)
    return mus, sigmas

def initial_hypothesis_mu(data, k):
    h = []
    minval = np.min(data)
    maxval = np.max(data)
    interval = (maxval - minval) / k
    h.append(minval)
    for i in xrange(1, k-1):
        h.append(minval + i * interval)
    h.append(maxval)
    return h

def compare_hypothesis(curr, old, epsilon=0.01):
    if not old:
        return True
    diff = 0
    for i in xrange(len(curr)):
        diff += curr[i] - old[i]
    return fabs(diff) > 0.01

def expected_value_point(point, mu, mus, sigma, sigmas):
    exp_num = prob_point_gauss(point, mu, sigma)
    exp_denom = 0
    for i in xrange(len(mus)):
        exp_denom += prob_point_gauss(point, mus[i], sigmas[i])
    return 0 if exp_denom == 0 else (exp_num / exp_denom)

def prob_point_gauss(point, mu, sigma):
    gauss_dist = norm(mu, sigma)
    return gauss_dist.pdf(point)

def calculate_expectation(k, data, mus, sigmas):
    exp_val = np.empty((k, data.size))
    for i in xrange(data.size):
        for j in xrange(k):
            exp_val[j][i] = expected_value_point(data[i], mus[j], mus, sigmas[j], sigmas)
    return exp_val

def calculate_hypothesis(k, data, expected_values):
    mus = []
    sigmas = []
    for j in xrange(k):
        mu_num = 0
        denom = 0
        sigma_num = 0
        for i in xrange(data.size):
            mu_num += expected_values[j][i] * data[i]
            denom += expected_values[j][i]
        mus.append(mu_num / denom)
        for i in xrange(data.size):
            sigma_num += expected_values[j][i] * (data[i] - mus[j])**2
        sigmas.append((sigma_num / denom)**0.5)
    return mus, sigmas

# prints means squared error
@timed
def test():
    difflist = []
    k = 3
    trials = 10
    for j in xrange(trials):
        data, mus = kgauss_with_mus(k, 100, dim=1, lower=-100, upper=100, sigma=3)  
        model_mus =  expectation_maximization(data[0], k, 3)
        model_mus.sort()
        mus.sort()
        print model_mus, mus
        difflist.append(tuple([(mus[i] - model_mus[i])**2 for i in xrange(k)]))
        print difflist[j]
    diffs = [0 for i in xrange(k)]
    for i in xrange(k):
        for j in xrange(10):
            diffs[i] += difflist[j][i]
        diffs[i] /= 10

def test_ppg():
    point, mean, sigma = randint(-100, 100), randint(-100, 100), randint(5, 20)
    print 'copyValueFromPython = ' + str(prob_point_gauss(point, mean, sigma)) + ';'
    print 'assertEquals(copyValueFromPython,ExpMax.probPoint(' + str(point) + ', ' + str(mean) + ', ' + str(sigma) +'),1e-4);'

def test_em():
    filename = '..\\Java\\temp2.txt'
    data = kgauss(2, 100, dim=1, lower=-100, upper=100, sigma=randint(5, 15))
    exportFile(filename, data)
    # print data[0]
    params = expectation_maximization(data[0], 2, [1, 1])
    print 'pythonParams = new double[][] {',
    for i in params:
        print '{',
        for j in i:
            print str(j) + ',',
        print '},',
    print '};'

def main():
    test_em()    

if __name__ == "__main__":
    main()

File: java/edu/cmu/ml/geoEM/ExpMax.java
package edu.cmu.ml.geoEM;
/* 
 * ExpMax.java
 * Description: A implementation of the expectation-maximization algorithm
 * for a mixture of k gaussians, with one-dimensional data
 * Author: Tom Shen
 * Date: 02/18/2013
 */

import java.io.IOException;
import java.util.*;
import org.apache.commons.math3.util.FastMath;
import org.apache.commons.math3.distribution.NormalDistribution;
import org.apache.commons.math3.stat.descriptive.rank.*;

public class ExpMax
{
    public double[] data;
    public int numDist;
    public int numPoints;
    public double[] means;
    public double[] sigmas;
    protected double epsilon = 0.01;

    
    public ExpMax(double[] data, int k) {
        this.data = data;
        numDist = k;
        numPoints = data.length;
        
        means = new double[numDist];
        double minVal = new Min().evaluate(data, 0, data.length);
        double maxVal = new Max().evaluate(data, 0, data.length);
        double interval = (maxVal - minVal) / (numDist - 1);
        for(int i = 0; i < numDist; i++)
            means[i] = minVal + i * interval;

        sigmas = new double[numDist];
        Arrays.fill(sigmas, 1.0);
    }

    protected double[][] calculateParameters() {
        double[] oldMeans = new double[numDist];
        double[] oldSigmas = new double[numDist];
        do {
            double[][] expectedValues = calculateExpectation();
            System.arraycopy(means, 0, oldMeans, 0, means.length);
            System.arraycopy(sigmas, 0, oldSigmas, 0, sigmas.length);
            calculateHypothesis(expectedValues);
        } while(compare(means, oldMeans) || compare(sigmas, oldSigmas));
        System.out.println("Model means:  " + Arrays.toString(means));
        System.out.println("Model sigmas: " + Arrays.toString(sigmas));
        return new double[][] {means, sigmas};
    }

    protected boolean compare(double[] curr, double[] old) {
        if(old == null)
            return true;
        double diff = 0.0;
        for(int i = 0; i < curr.length; i++)
            diff += FastMath.pow(curr[i] - old[i], 2.0);
        return FastMath.pow(diff, 0.5) > epsilon;
    }

    protected double[][] calculateExpectation() {
        double[][] expectedValues = new double[numDist][numPoints];
        for(int i = 0; i < numDist; i++)
            for(int j = 0; j < numPoints; j++)
                expectedValues[i][j] = expectedValuePoint(data[j], i);
        return expectedValues;
    }

    protected double expectedValuePoint(double point, int currDist) {
        double probCurrDist = probPoint(point, means[currDist], sigmas[currDist]);
        double probAllDist = 0;
        for(int i = 0; i < numDist; i++)
            probAllDist += probPoint(point, means[i], sigmas[i]);
        if(probAllDist == 0)
            return 0;
        return probCurrDist / probAllDist;
    }

    protected static double probPoint(double point, double mean, double sigma) {
        NormalDistribution dist = new NormalDistribution(mean, sigma);
        return dist.density(point);
    }
    
    protected void calculateHypothesis(double[][] expectedValues) {
        for(int i = 0; i < numDist; i++) {
            double totalExp = 0;
            means[i] = 0;
            sigmas[i] = 0;
            for(int j = 0; j < numPoints; j++) {
                means[i] += expectedValues[i][j] * data[j];
                totalExp += expectedValues[i][j];
            }
            means[i] /= totalExp;
            for(int j = 0; j < numPoints; j++) {
                sigmas[i] += expectedValues[i][j]
                             * FastMath.pow((data[j] - means[i]), 2.0);
            }
            sigmas[i] = FastMath.pow(sigmas[i] / totalExp, 0.5);
        }
    }
    public static void main(String args[]) throws IOException {
        int k = 3;
        Util.exportFile("temp.txt", KGauss.kgauss(k, 100, 1, -100, 100, 0.1));
        double[] data = Util.importFile("temp.txt")[0];
        ExpMax em = new ExpMax(data, k);
        em.calculateParameters();
    }
}